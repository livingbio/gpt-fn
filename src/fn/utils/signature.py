import inspect
from typing import Any, Callable

import pydantic

from .pydantic_parser import PydanticParser


def clean_docstring(docstring: str) -> str:
    """Clean up docstring before sending to OpenAI."""
    output = []
    for line in docstring.split("\n"):
        output.append(line.strip())

    return "\n".join(output)


class FunctionSignature:
    """A helper class to parse function signature and generate instruction for prompt."""

    def __init__(self, fn: Callable[..., Any]):
        self.fn = fn
        self.sig: inspect.Signature = inspect.signature(fn)
        return_annotation = self.sig.return_annotation

        assert return_annotation is not inspect.Signature.empty, f"Function {fn.__name__} must have return annotation"

        class UnpackModel(pydantic.BaseModel):
            ret: return_annotation  # type: ignore[valid-type]

        self.parser = PydanticParser[UnpackModel](pydantic_model=UnpackModel)

    def function_line(self) -> str:
        # NOTE: return type is instrcutive by parser
        f = str(self.sig).split("->")[0].strip()

        return f"def {self.fn.__name__}{f}:"

    def description(self) -> str:
        desc = inspect.cleandoc(self.fn.__doc__ or "")

        return desc

    def call_line(self, *args: Any, **kwargs: Any) -> str:
        # Bind the provided arguments to the function signature
        bound_args = self.sig.bind(*args, **kwargs)
        bound_args.apply_defaults()

        # Build input binds
        input_binds: list[str] = []
        for k, v in bound_args.arguments.items():
            input_binds.append(f"{k}={repr(v)}")

        return f"{self.fn.__name__}({', '.join(input_binds)})"

    def parse(self, text: str) -> Any:
        return self.parser.parse(text).ret

    def instruction(self) -> str:
        return clean_docstring(
            f"""You are now the following python function:
                    ```
                    # {self.description()}
                    {self.function_line()}
                    ```
                    Only respond with your `return` value.
                    {self.parser.get_format_instructions()}
                    """
        )
