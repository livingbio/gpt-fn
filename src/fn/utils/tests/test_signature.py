from typing import Any, Callable

import pytest
from pydantic import BaseModel
from syrupy.assertion import SnapshotAssertion

from ..signature import FunctionSignature


def add(a: int, b: int) -> int:  # type: ignore[empty-body]
    """Add two numbers."""


class Person(BaseModel):
    """Person model."""

    name: str
    age: int


def fake_person(count: int) -> Person:  # type: ignore[empty-body]
    """generate fake person."""


def no_return_annoation(a: int, b: int):  # type: ignore[no-untyped-def]
    """Add two numbers."""


@pytest.mark.parametrize("func", [add, fake_person])
def test_function_signature(func: Callable[[Any], Any], snapshot: SnapshotAssertion) -> None:
    sig = FunctionSignature(func)
    assert snapshot == sig.instruction()


def test_function_signature_without_return_annoations() -> None:
    with pytest.raises(AssertionError):
        FunctionSignature(no_return_annoation)
