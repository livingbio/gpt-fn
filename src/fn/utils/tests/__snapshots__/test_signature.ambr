# serializer version: 1
# name: test_function_signature[add]
  '''
  You are now the following python function:
  ```
  # Add two numbers.
  def add(a: int, b: int):
  ```
  Only respond with your `return` value.
  The output should be formatted as a JSON instance that conforms to the JSON schema below.
  
  As an example, for the schema {"properties": {"foo": {"title": "Foo", "description": "a list of strings", "type": "array", "items": {"type": "string"}}}, "required": ["foo"]}}
  the object {"foo": ["bar", "baz"]} is a well-formatted instance of the schema. The object {"properties": {"foo": ["bar", "baz"]}} is not well-formatted.
  
  Here is the output schema:
  ```
  {"properties": {"ret": {"title": "Ret", "type": "integer"}}, "required": ["ret"]}
  ```
  
  '''
# ---
# name: test_function_signature[fake_person]
  '''
  You are now the following python function:
  ```
  # generate fake person.
  def fake_person(count: int):
  ```
  Only respond with your `return` value.
  The output should be formatted as a JSON instance that conforms to the JSON schema below.
  
  As an example, for the schema {"properties": {"foo": {"title": "Foo", "description": "a list of strings", "type": "array", "items": {"type": "string"}}}, "required": ["foo"]}}
  the object {"foo": ["bar", "baz"]} is a well-formatted instance of the schema. The object {"properties": {"foo": ["bar", "baz"]}} is not well-formatted.
  
  Here is the output schema:
  ```
  {"properties": {"ret": {"$ref": "#/definitions/Person"}}, "required": ["ret"], "definitions": {"Person": {"title": "Person", "description": "Person model.", "type": "object", "properties": {"name": {"title": "Name", "type": "string"}, "age": {"title": "Age", "type": "integer"}}, "required": ["name", "age"]}}}
  ```
  
  '''
# ---
